[Merged by Bors] - bevy_app: add tracing event with `tracy.frame_mark`

jakobhellermann opened the pull request:
Currently `tracy` interprets the entire trace as one frame because the marker for frames isn't being recorded.

~~When an event with `tracy.trace_marker=true` is recorded, `tracing-tracy` will mark the frame as finished:
<https://github.com/nagisa/rust_tracy_client/blob/aa0b96b2ae2aee6f3831994f9a3a8c29bcc85fae/tracing-tracy/src/lib.rs#L240>~~

~~Unfortunately this leads to~~
```rs
INFO bevy_app:frame: bevy_app::app: finished frame tracy.frame_mark=true
```
~~being printed every frame (we can't use DEBUG because bevy_log sets `max_release_level_info`.~~

Instead of emitting an event that gets logged every frame, we can depend on tracy-client itself and call `finish_continuous_frame!();`

superdump reviewed (changes requested):
Did you try using a tracing event with the frame marker but at the debug level or so? Does it just get filtered out before it gets sent to Tracy? I think that aspect was maybe why I didn't make a PR.

Perhaps if that is a negative side effect of using the tracing frame marker API, we could make a contribution to tracing-tracy to allow directly calling finish continuous frame for when an event or message or so is undesirable.

jakobhellermann commented:
> Is it a good idea to depend on both tracy_client and tracing_tracy?

`tracing_tracy` uses `tracy_client` will call `finish_continous_frame!` when an event with `tracy.frame_mark` is recorded: <https://github.com/nagisa/rust_tracy_client/blob/aa0b96b2ae2aee6f3831994f9a3a8c29bcc85fae/tracing-tracy/src/lib.rs#L240>
So this PR is doing exactly the same, just without the event.

> Did you try using a tracing event with the frame marker but at the debug level or so?

I tried but that didn't work in release mode as `bevy_log` enabled the `max_release_level_info` feature on `tracing-subscriber`.

jakobhellermann commented on crates/bevy_app/src/app.rs:
```diff
@@ -114,6 +114,9 @@ impl App {
         }
+
+        #[cfg(feature = "tracing-tracy")]
+        tracy_client::finish_continuous_frame!();
```
So maybe the correct position would be after this `if`?
<https://github.com/bevyengine/bevy/blob/207ebde0200d763725569feb8547cd6a5e18c327/crates/bevy_render/src/renderer/mod.rs#L63-L65>

The `take_surface_texture` moves the `SurfaceTexture` out of the `Arc`, and when it is dropped wgpu calls `Context::surface_texture_discard`.

jakobhellermann replied:
I wonder if tracy can make sense of the instrumentation when we move to actual pipelined rendering.

superdump replied:
I don’t know how it works but I think it’s something like traciing the command submissions from the CPU to the GPU allowing the association to be made.

robtfm commented:
just to note an alternative to using another import, we could also suppress the info logging for tracy events:

```diff
+++ b/crates/bevy_log/src/lib.rs
@@ -159,6 +159,11 @@ impl Plugin for LogPlugin {
             let tracy_layer = tracing_tracy::TracyLayer::new();

             let fmt_layer = tracing_subscriber::fmt::Layer::default();
+            #[cfg(feature = "tracing-tracy")]
+            let fmt_layer = fmt_layer.with_filter(
+                tracing_subscriber::filter::Targets::new().with_target("tracy", Level::ERROR),
+            );
+
             let subscriber = subscriber.with(fmt_layer);

             #[cfg(feature = "tracing-chrome")]
```

superdump commented:
@robtfm oh that’s a good find. I guess warnings would be ok

superdump commented:
@jakobhellermann  what do you think of @robtfm ’s suggestion?

jakobhellermann commented:
I switched to @robtfm's suggestion to use the event but exclude it from the `fmt_layer` and moved the marker to right after the surface is presented and the `SurfaceTexture` dropped.

mockersf commented:
bors r+
